server/athenian/api/controllers/search_controller/search_prs.py

```diff
@@ -220,3 +240,137 @@ async def _search_pr_digests(
+    ) -> Optional[_OrderByMetrics]:
+        metrics = [expr.field for expr in order_by if expr.field in cls.FIELDS]
+        assert metrics
+        min_times = np.array(
+            [search_filter.time_from.replace(tzinfo=None)], dtype="datetime64[ns]",
+        )
+        max_times = np.array([search_filter.time_to.replace(tzinfo=None)], dtype="datetime64[ns]")
+        calc_ensemble = PullRequestMetricCalculatorEnsemble(
+            *metrics, quantiles=(0, 1), quantile_stride=DEFAULT_QUANTILE_STRIDE,
+        )
+        groups = np.full((1, len(pr_facts)), True, bool)
+        calc_ensemble(pr_facts, min_times, max_times, groups)
+        return cls(calc_ensemble)
+
+    def parse_expression(
+        self,
+        expr: SearchPullRequestsOrderByExpression,
+    ) -> tuple[npt.NDArray, npt.NDArray[int]]:
+        """Parse an expression and return a tuple with order keys and discard indexes.
+
+        `order_keys` is an array with the same length as `pr_facts` with the sorting key for
+        this expression for each pr.
+        `discard` is an array with the indexes of element in `pr_facts` to drop from end result
+        due to expression.
+
+        """
+        # first calculator is taken, calculators using split() are not yet supported
+        calc = self._calc_ensemble[expr.field][0]
+        values = calc.peek[0]
+        assert len(values)
+
+        discard = np.array([], dtype=int)
+        nulls, null_replacement = self._filter_nulls(values, expr, calc)
+        if len(nulls):
+            if expr.exclude_nulls:
+                # no need to replace nulls with real values, their ordering can be any
+                # as they will not be kept
+                discard = nulls
+            else:
+                values = values.copy()
+                values[nulls] = null_replacement
+
+        if expr.direction == OrderByDirection.DESCENDING.value:
+            values = -values
+        return values, discard
+
+    @classmethod
+    def _filter_nulls(
+        cls,
+        ar: npt.NDArray,
+        expr: SearchPullRequestsOrderByExpression,
+        calc: MetricCalculator,
+    ) -> tuple[npt.NDArray, int | float | np.timedelta64]:
+        """Search for "nulls" in array depending on dtype and also return the replacement value.
+
+        Replacement is needed to invert null values when the direction is descending.
+        Replacement value is max or min depending on `direction`, `nulls_first` in the expression.
+        """
+        if expr.direction == OrderByDirection.DESCENDING.value:
+            # with descending order I want a huge null if they should be first in the result
+            nullmax = expr.nulls_first
+        else:
+            # with ascending order I want a huge null if they should last in the result
+            nullmax = not expr.nulls_first
+
+        null_replacement: int | float | np.timedelta64
+
+        if ar.dtype == np.dtype("timedelta64[s]"):
```

Comment:
We should check against `Calculator.nan` like this:
```python
if calc.has_nan:
    nulls = ar != ar
else:
    nulls = ar == calc.nan
```

Reply:
Regarding the replacements, the whole approach should be different (and way simpler!):

```python
notnulls = ~nulls
notnull_indexes = np.flatnonzero(notnulls)[np.argsort(ar[notnull]s)]
null_indexes = np.flatnonzero(nulls)
# then decide depending on `nullmax` whether to set indexes = np.concatenate([notnull_indexes, null_indexes]) or vice versa
```

The goal is to avoid all the custom code.

Reply:
I've replaced `nulls` building with the suggested code.
Regarding replacements, I understand this code generates indexes to order by a single expression,  can it be adapted for multiple expressions ?
My approach generates a list of sorting keys and pass it to `lexsort` so I need to convert nulls to some value to control their position for each sorting keys.
The ugly part is ofc the switch on calc.dtype to get replacements

Reply:
You should ditch lexsort in favor of several regular `np.sort`-s. The performance will be similar if not better because lexsort is slow. Even it becomes worse, it's still worth because the current code is too complex and tricky.

Reply:
To elaborate, `np.sort(kind="stable")` for each column in the opposite direction with nulls appended or prepended.

Reply:
I mean this:
```python
state = np.arange(len())
for expr in reversed(expressions):
    ar = ...  # column
    ar = ar[state]
    if calc.has_nan:
        nulls = ar != ar
    else:
        nulls = ar == calc.nan
    state_notnull = state[~nulls][np.argsort(ar[~nulls], kind="stable")]
    state_parts = [state[nulls], state_notnull]
    if expr.nulls_last:
        state_parts = state_parts[::-1]
    state = np.concatenate(state_parts)
```

Reply:
nice, this is indeed simpler